### PR TITLE
fix some bugs in diagonal subtyping

### LINKED ISSUE #24166: Subtyping is not reflexive and not transitive with triangular rule
In julia 0.6.0 subtyping is not reflexive for some examples.
```julia
julia> (Tuple{T, T} where Int<:T<:Int) <: (Tuple{T, T} where Int<:T<:Int)
false

julia> (Tuple{T, T} where T>:Int) <: (Tuple{T, T} where T>:Int)
false
```
It seems to me that there is a problem with lower bounds in triangular dispatched cases, because these examples work fine:
```julia
julia> (Tuple{T, T} where T>:Number) <: (Tuple{T, T} where T>:Number)
true

julia> (Tuple{T, T} where T<:Int) <: (Tuple{T, T} where T<:Int)
true
```

Also, transitivity is broken:
```julia
julia> T1 = Tuple{Number, Number, Ref{Number}}
Tuple{Number,Number,Ref{Number}}

julia> T2 = Tuple{T, T, Ref{T}} where T
Tuple{T,T,Ref{T}} where T

julia> T3 = Tuple{S, S, Ref{Q} where Q} where S
Tuple{S,S,Ref{Q} where Q} where S

julia> T1 <: T2
true

julia> T2 <: T3
true

julia> T1 <: T3
false
```
Although, to my understanding, `T1 <: T2 ==  true` and `T1 <: T3 ==  false` are correct, but `T2 <: T3 == true` is not. So it's not really transitivity, but rather some problem with `T2` on the left.

Finally, this looks weird:
```julia
julia> (Tuple{T, T} where Number<:T<:Number) <: Tuple{Number, Number}
true

julia> Tuple{Number, Number} <: (Tuple{T, T} where Number<:T<:Number)
true

julia> Vector{Tuple{T, T} where Number<:T<:Number} <: Vector{Tuple{Number, Number}}
false
```
I don't see why vectors should not be in subtype relation.

Here is the version info:
```
Julia Version 0.6.0
Commit 903644385b (2017-06-19 13:05 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i5-4200U CPU @ 1.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
```
@JeffBezanson

### LINKED ISSUE #31824: tuple splatting bug with simple program
version information:
```
Julia Version 1.0.3
Commit 099e826241 (2018-12-18 01:34 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, broadwell)
Environment:
  JULIA_EDITOR = "C:\JuliaPro-1.0.3.1\app-1.32.2\atom.exe" -a
  JULIA_NUM_THREADS = 16
  JULIA_PKG_SERVER = https://pkg.juliacomputing.com/
  JULIA_PKG_TOKEN_PATH = C:\Users\bguenter\.juliapro\token.toml
```
Smallest function that produces the bug

```
julia> function test(input...)
       b::NTuple = input
       end
test (generic function with 1 method)
```

How to produce the bug
```
julia> test(1,2,3)

Unreachable reached at 000000001DBCAA98

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (i
n their entirety). Thanks.
Exception: EXCEPTION_ILLEGAL_INSTRUCTION at 0x1dbcaa98 -- test at .\none:2
in expression starting at no file:0
test at .\none:2
jl_fptr_trampoline at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:1831
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2184do_call at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:324
eval_value at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:430
eval_stmt_value at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:363 [in
lined]
eval_body at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:678
jl_interpret_toplevel_thunk_callback at /home/Administrator/buildbot/worker/package_win64/build/src\
interpreter.c:806
unknown function (ip: FFFFFFFFFFFFFFFE)
unknown function (ip: 00000000119D37FF)
unknown function (ip: FFFFFFFFFFFFFFFF)
jl_toplevel_eval_flex at /home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:805
jl_toplevel_eval_in at /home/Administrator/buildbot/worker/package_win64/build/src\builtins.c:622
eval at .\boot.jl:319 [inlined]
repleval at C:\Users\bguenter\.juliapro\JuliaPro_v1.0.3.1\packages\Atom\v2iqN\src\repl.jl:139
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2184#164 at C:\Users\bguenter\.juliapro\JuliaPro_v1.0.3.1\packages\Atom\v2iqN\src\repl.jl:161with_logstate at .\logging.jl:395
with_logger at .\logging.jl:491 [inlined]
evalrepl at C:\Users\bguenter\.juliapro\JuliaPro_v1.0.3.1\packages\Atom\v2iqN\src\repl.jl:152
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2184do_call at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:324
eval_value at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:430
eval_stmt_value at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:363 [in
lined]
eval_body at /home/Administrator/buildbot/worker/package_win64/build/src\interpreter.c:678
jl_interpret_toplevel_thunk_callback at /home/Administrator/buildbot/worker/package_win64/build/src\
interpreter.c:806
unknown function (ip: FFFFFFFFFFFFFFFE)
unknown function (ip: 00000000119D348F)
unknown function (ip: FFFFFFFFFFFFFFFF)
jl_toplevel_eval_flex at /home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:805
jl_toplevel_eval_in at /home/Administrator/buildbot/worker/package_win64/build/src\builtins.c:622
eval at .\boot.jl:319
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2184eval_user_input at C:\Users\julia\AppData\Local\Julia-1.0.3\share\julia\stdlib\v1.0\REPL\src\REPL.jl
:85
macro expansion at C:\Users\julia\AppData\Local\Julia-1.0.3\share\julia\stdlib\v1.0\REPL\src\REPL.jl
:117 [inlined]
#28 at .\task.jl:259
jl_fptr_trampoline at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:1831
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2184jl_apply at /home/Administrator/buildbot/worker/package_win64/build/src\julia.h:1537 [inlined]
start_task at /home/Administrator/buildbot/worker/package_win64/build/src\task.c:268
Allocations: 30636752 (Pool: 30631301; Big: 5451); GC: 69

Julia has exited. Press Enter to start a new session.
```

### PR DESCRIPTION
**Pull request by JeffBezanson**
fixes #31824, fixes #24166

This is slightly breaking, since `convert(NTuple, (1,""))` used to work but now gives a MethodError. I'm convinced it's correct though, since NTuple means "homogeneous tuple".

### CODE REVIEW

File: test/tuple.jl [hunk 1]
```diff
@@ -28,11 +28,8 @@ end
     @test convert(Tuple{Int, Int, Float64}, (1, 2, 3)) === (1, 2, 3.0)
 
     @test convert(Tuple{Float64, Int, UInt8}, (1.0, 2, 0x3)) === (1.0, 2, 0x3)
-    @test convert(NTuple, (1.0, 2, 0x3)) === (1.0, 2, 0x3)
```
Comment (Keno) [inline test/tuple.jl]: change to `NTuple{N, Real} where N` or something similar?

File: test/tuple.jl [hunk 2]
```diff
@@ -28,11 +28,8 @@ end
...
     @test convert(Tuple{Vararg{Int}}, (1.0, 2, 0x3)) === (1, 2, 3)
     @test convert(Tuple{Int, Vararg{Int}}, (1.0, 2, 0x3)) === (1, 2, 3)
-    @test convert(Tuple{Vararg{T}} where T<:Integer, (1.0, 2, 0x3)) === (1, 2, 0x3)
```
Comment (Keno) [inline test/tuple.jl]: `Tuple{Vararg{Integer}}`?

Comment (JeffBezanson) [reply test/tuple.jl]: Good point. In fact we need a new method to handle that now.

Comment (JeffBezanson) [reply test/tuple.jl]: So, one downside to making the element type inside `Vararg` diagonal is that `Tuple{Vararg{Integer}}` doesn't match `Tuple{Vararg{T}} where T`. In fact I can't find a type that includes all such vararg types without also including `NTuple`, which you can't sensibly convert to. AFAICT, the best we can do is adding a type assertion to the conversion method to make sure the result is valid.

Comment (Keno) [reply test/tuple.jl]: Arguably, I'm not convinced the following normalization is correct:
```
julia> Tuple{Vararg{T where T<:S}} where S
Tuple{Vararg{S,N} where N} where S
```

Comment (JeffBezanson) [reply test/tuple.jl]: Fair point; I would put all the blame on the diagonal rule however --- it's too structure-based, so potentially all sorts of rearrangements get in its way.

File: test/tuple.jl [hunk 3]
```diff
@@ -53,6 +52,11 @@ end
     @test_throws MethodError convert(Tuple{Int, Int, Int}, (1, 2))
     # issue #26589
     @test_throws MethodError convert(NTuple{4}, (1.0,2.0,3.0,4.0,5.0))
+    # issue #31824
+    # there is no generic way to convert an arbitrary tuple to a homogeneous tuple
+    @test_throws MethodError convert(NTuple, (1, 1.0))
+    @test_throws MethodError convert(Tuple{Vararg{T}} where T<:Integer, (1.0, 2, 0x3)) === (1, 2, 0x3)
```
Comment (vtjnash) [inline test/tuple.jl]:
```suggestion
    let T = Tuple{Vararg{T}} where T<:Integer, v = (1.0, 2, 0x3)
        @test_throws MethodError(convert, (T, v)) convert(T, v)
    end
```

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by maleadt** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/31833#issuecomment-565201623):
```
NanosoldierError: failed to run tests against primary commit: UndefVarError: prnumber not defined
```
Logs and partial data can be found [here](https://github.com/maleadt/BasePkgEvalReports/blob/7b088c9eb5ee332bbe2b3651cce2927f7074a0fc/pkgeval-d934db0_vs_7090b23/report.md)
cc @maleadt

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by maleadt** [comment]
> `NanosoldierError: failed to run tests against primary commit: UndefVarError: prnumber not defined`

Ah, that's a bug. I'll fix it and re-trigger.

**Comment by maleadt** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your test job](https://github.com/JuliaLang/julia/pull/31833#issuecomment-565934549) has completed - possible new issues were detected. A full report can be found [here](https://github.com/maleadt/BasePkgEvalReports/blob/5f202daa59113dfff0e71fe456459f539eeb33cd/pkgeval-d934db0_vs_a6bf74a/report.md). cc @maleadt

**Comment by JeffBezanson** [comment]
Ok, this has some legit regressions. Can wait for 1.4.1 to see if those can be avoided.

**Pull request closed without merging**